File: bintree_java/src/binarytree/BinNode.java
package binarytree;



public class BinNode< TKey extends Number & Comparable<? super TKey>,
                      TNode extends BinNode<TKey,TNode>
                    >
{


    public TKey key = null;
    public TNode left = null;
    public TNode right = null;



    public BinNode() {
    }



    public BinNode(TKey key) {
        this.key = key;
    }



    @SuppressWarnings("unchecked")
    public TNode cloneBranch() {
        var theClone = new BinNode<TKey, TNode>(this.key);

        if (null != this.left)
            theClone.left = this.left.cloneBranch();

        if (null != this.right)
            theClone.right = this.right.cloneBranch();

        return (TNode) theClone;
    }
}

File: bintree_csharp/binarytree/BinNode.cs
using System;


namespace my.binarytree
{
    class BinNode<TKey, TNode>
        where TKey : IComparable where TNode : BinNode<TKey, TNode>
    {
        public TKey Key = default;
        public TNode Left = null;
        public TNode Right = null;



        public BinNode(TKey key = default)
        {
            this.Key = key;
        }
    }
}
